### ISSUE TITLE
Medusa needs a command line/config option to select the testing mode

### ISSUE THREAD

**Issue by ggrieco-tob** [crytic/medusa#146]
With the upcoming optimization mode #71 and new assertions checks with other panics #142, it could be a good idea to rethink the usage of different configuration options, which results in the following structure:

```json
"assertionTesting": {
	"enabled": false,
	"testViewMethods": false
},
"propertyTesting": {
	"enabled": false,
	"testPrefixes": [
		"fuzz_"
	]
},
"optimizationTesting": {
	"enabled": true,
	"testPrefixes": [
		"optimize_"
	]
}
``` 
Some of the options are repeated and the user is forced to go over each one to enable or disabled. Also, it is likely that enabling two or more at the same time will never be used in practice. To overcome this in echidna, the parameter used is `testing-mode` which can be set to a string: `property`, `assertion`, `optimization` (and a few others). It could be a good idea to implement this in medusa. In particular, to allow switching between modes quite easily.

**Comment by 0xalpharush** [comment]
I think the requirement to understand what each of these modes does and which to select is unfriendly to new users. The solution proposed here https://github.com/crytic/echidna/issues/941#issuecomment-1532674974 makes sense, but I would offer the following as a solution.

Generate a folder structure for each mode and a stub of a fuzz function (maybe with a concise explanation). This would make it clear what mode is being used, how to use it,  and provide templates for how to quickly get started.
```
fuzz/ 
     assertion/
     property/
     optimization/
```
`assertion/` would contain a simple contract with a function that uses `assert`, property mode would have a function that returns a boolean and is named `invarinat_*`, etc.

Then, someone could ideally just run `medusa fuzz --target fuzz` (make default?) and all of the fuzz cases would run in their respective mode.

**Comment by Xenomega** [comment]
The initial suggestion makes sense to me on a CLI level. 

With the comment above @0xalpharush, I may misunderstand intent so correct me if I got it wrong. My initial thoughts are that  if we create different folders that target different testing methodologies (for each project created in medusa), the structure may get weird if you use a mix of property/assertion testing in your codebase and don't want to recompile/copy-paste/re-import your files in both directory structures. Or it may compile some targets twice then and get a bit weird if you actually do want to use both testing methodologies. And if we don't let you use more than one at once, you'll lose efficiency running two instances of the fuzzer.

If the idea is to basically have tests folders in your smart contract folder for each testing methodology, maybe its best to leave the config open-ended/flexible, and introduce your idea as part of a "front end testing harness" we discussed on our last sync call? We could just make it create this folder structure for you by default in that harness, and set up the config to point to all the right areas?

TLDR; I figured the backend of the fuzzer should allow a lot of customizability for different use cases we/others may have in the future, while the front end testing harness we'd tack on later could build specific project structures that make fuzzing with medusa more intuitive.

If this was a total miss or you think my reasoning maybe is weak somewhere though, let me know. I'm open to the idea of re-structuring some of this if we think it would work well with all of our incentives.

**Comment by 0xPhaze** [comment]
The best way that I can imagine the different modes to work for me, would be:
- The same configurable prefix (`test` or `medusa_`) for all fuzzed functions
- Testing mode is only determined via the return value
  - No return value => assertion mode: only assertions trigger a failed test-case
  - Return `bool` => property testing: failure if the function does not return `true`
  - Return `uint`/`int` => optimization mode: maximize the return value
- Assertion mode is not really a separate mode, i.e. any `assert(false)` should trigger a panic and be reported

This way we do not impose any constraints on folder/file structure and tests from foundry can also easily be ported over.

**Comment by 0xalpharush** [comment]
I agree the directories structure would probably be over restrictive, but I like the stub idea and will have to think about it more.. 

I think @0xPhaze's suggestion would leave the only thing to be configured the prefix of property and optimization tests and whether to fuzz view/ pure functions (which I think should be enabled by default for stateless fuzzing). We could also add a way to disable/ customize assertion failures and open it up to encompass something like overflows if desired.

**Comment by anishnaik** [comment]
When we merge assertion and property mode #191 , I believe that this issue could be indirectly solved. Personally, I think the way we educate the ecosystem on various ways to fuzz test is a bit flawed. At the end of the day we are testing properties and what mode is selected can be quite confusing to the user. So both modes should be provided out-of-box and how optimization mode is configuration, branded, and enabled should be handled in the PR for #191 .
